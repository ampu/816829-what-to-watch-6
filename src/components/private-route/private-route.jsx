import React, {useCallback} from 'react';
import PropTypes from 'prop-types';
import {Route, Redirect} from 'react-router-dom';
import {connect} from 'react-redux';

import {MainPath} from '../../constants/paths';
import {OperationStatus, OPERATION_STATUSES} from '../../constants/operation-status';
import SpinnerLoading from '../spinner-loading/spinner-loading';

const PrivateRoute = ({exact = true, path, render, loginStatus}) => {

  const doRender = useCallback((routeProps) => {
    if (loginStatus === OperationStatus.PENDING) {
      return <SpinnerLoading/>;
    }
    return loginStatus === OperationStatus.RESOLVED
      ? render(routeProps)
      : <Redirect to={MainPath.SIGN_IN}/>;
  }, [render, loginStatus]);

  return <Route exact={exact} path={path} render={doRender}/>;
};

PrivateRoute.propTypes = {
  exact: PropTypes.bool,
  path: PropTypes.string.isRequired,
  render: PropTypes.func.isRequired,
  loginStatus: PropTypes.oneOf(OPERATION_STATUSES),
};

const mapStateToProps = (state) => ({
  loginStatus: state.loginStatus,
});

export {PrivateRoute};
export default connect(mapStateToProps)(PrivateRoute);
