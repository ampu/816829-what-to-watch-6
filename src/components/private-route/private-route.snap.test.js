import React from 'react';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from 'history';

import {render} from '@testing-library/react';

import OperationStatus from '../../constants/operation-status';

import {PrivateRoute} from './private-route';

describe(`PrivateRoute component`, () => {
  it(`Should render 'SpinnerLoading' when login status is pending`, () => {
    const history = createMemoryHistory();
    const doRender = jest.fn();

    const {container} = render((
      <Router history={history}>
        <PrivateRoute loginStatus={OperationStatus.PENDING} path="/" render={doRender}/>
      </Router>
    ));

    expect(container).toMatchSnapshot();
  });
});
