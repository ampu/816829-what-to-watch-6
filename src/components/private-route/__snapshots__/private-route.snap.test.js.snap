// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute component Should render 'SpinnerLoading' when login status is pending 1`] = `
<div>
  <div
    class="container container--centered"
  >
    <div
      class="spinner-loading"
    >
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
    </div>
  </div>
</div>
`;
